### PR TITLE
7903128: Overhaul jextract README

### PR DESCRIPTION
**Pull request by mcimadamore**
I've improved the README file, by putting the expanding the build section, and putting it more front and center.
I have also streamlined the remaining contents a little, and removed the parts that were describing the foreign API (not really needed here).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [CODETOOLS-7903128](https://bugs.openjdk.java.net/browse/CODETOOLS-7903128): Overhaul jextract README

### Reviewers
 * [Athijegannathan Sundararajan](https://openjdk.java.net/census#sundar) (@sundararajana - no project role) ⚠️ Review applies to b2b97d165dda58054bfa4ffb633773f7a1040272
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - no project role) ⚠️ Review applies to a64355175f1a8820f004fca3b2adbd79d788a8fa

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jextract pull/3/head:pull/3` \
`$ git checkout pull/3`

Update a local copy of the PR: \
`$ git checkout pull/3` \
`$ git pull https://git.openjdk.java.net/jextract pull/3/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3`

View PR using the GUI difftool: \
`$ git pr show -t 3`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jextract/pull/3.diff">https://git.openjdk.java.net/jextract/pull/3.diff</a>

</details>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,9 +108,11 @@ public static double distance ( MemorySegment x0) {
...
 
-The `jextract` tool includes several customization options. Users can select what in which package the generated code should be emitted, and what the name of the main extracted class should be. To allow for symbol filtering, jextract can generate a *dump* of all the symbols encountered in an header file; this dump can be manipulated, and then used as an argument file (using the `@argfile` syntax also available in other JDK tools) to e.g. generate bindings only for a *subset* of symbols seen by `jextract`. For instance, if we run `jextract` with as follows:
+#### Filtering symbols
+
+The `jextract` tool includes several customization options. Users can select what in which package the generated code should be emitted, and what the name of the main extracted class should be. To allow for symbol filtering, `jextract` can generate a *dump* of all the symbols encountered in an header file; this dump can be manipulated, and then used as an argument file (using the `@argfile` syntax also available in other JDK tools) to e.g. generate bindings only for a *subset* of symbols seen by `jextract`. For instance, if we run `jextract` with as follows:
```
Comment (JornVernee) [inline README.md:115]:
```suggestion
The `jextract` tool includes several customization options. Users can select in which package the generated code should be emitted, and what the name of the main extracted class should be. To allow for symbol filtering, `jextract` can generate a *dump* of all the symbols encountered in an header file; this dump can be manipulated, and then used as an argument file (using the `@argfile` syntax also available in other JDK tools) to e.g. generate bindings only for a *subset* of symbols seen by `jextract`. For instance, if we run `jextract` with as follows:
```

File: README.md [hunk 2]
```diff
@@ -1,65 +1,47 @@
...
-
-MemorySegment segment = MemorySegment.allocateNative(POINT_2D, ReosurceScope.newImplicitScope());
-xHandle.set(segment, 3d);
-yHandle.set(segment, 4d);
+```sh
+$ sh ./gradlew -Pjdk18_home=<jdk18_home_dir> -Plibclang_home=<libclang_dir> clean verify
 ```
 
-Furthermore, clients can use the `CLinker` API to create a so called *downcall* method handle, that is, a method handle which targets a native function, directly:
+After building, there should be a new `jextract` folder under `build` (the contents of this folder might vary depending on the platform):
```
Comment (sundararajana) [inline README.md:15]: Folder name is jextract.app on Mac. i.e., even folder name depends on platform

**Review by sundararajana — APPROVED** [review]
LGTM

File: README.md (hunk 1 above)
Comment (JornVernee) [reply README.md:115]: I think this should also either list or reference the corresponding command line options. i.e. `-t`, `--header-class-name`. For filtering the example below is good enough

**Review by JornVernee — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -1,65 +1,47 @@
...
 
-The Java SE 18 API defines an [incubating API](https://openjdk.java.net/jeps/419) to:
+### Getting started
 
-* manipulate foreign memory, that is memory which resides *outside* the Java heap; and
+`jextract` depends on the [C libclang API](https://clang.llvm.org/doxygen/group__CINDEX.html). To build the jextract sources, the easiest option is to download LLVM binaries for your platform, which can be found [here](https://releases.llvm.org/download.html) (a version <= 9 is required). Both the `jextract` tool and the bindings it generates depend heavily on the [Foreign Function & Memory API](https://openjdk.java.net/jeps/419), so a suitable [jdk 18 distribution](https://jdk.java.net/18/) is also required.
```
Comment (JornVernee) [inline README.md:7]: should be `version >= 9` right?

File: README.md [hunk 4]
```diff
@@ -126,9 +108,26 @@ public static double distance ( MemorySegment x0) {
...
+The `jextract` tool includes several customization options. Users can select in which package the generated code should be emitted, and what the name of the main extracted class should be. A complete list of all the supported options is given below:
+
+| Option                                                       | Meaning                                                      |
+| :----------------------------------------------------------- | ------------------------------------------------------------ |
+| `--header-class-name <String>`                               | specify the name the main header class                       |
```
Comment (JornVernee) [inline README.md:119]:
```suggestion
| `--header-class-name <String>`                               | specify the name of the main header class                       |
```

File: README.md [hunk 5]
```diff
@@ -126,9 +108,26 @@ public static double distance ( MemorySegment x0) {
...
+| `-t, --target-package <String>`                              | specify target package for the generated bindings            |
+| `-I <String>`                                                | specify include files path for the clang parser              |
+| `-l <String>`                                                | specify a library that will be loaded by the generated bindings |
+| `-d <String>`                                                | specify where to place generated files                       |
+| `--source`                                                   | generate java sources instead of classfiles                  |
+| `--dump-includes <String>`                                   | dump included symbols into specified file (see below)        |
+| `--include-[function,macro,struct,union,typedef,var]<String>` | Include a symbol of the given name and kind in the generated bindings (see below). When one of these options is specified, any symbols that is not matched by any specified filters is omitted from the generated bindings. |
```
Comment (JornVernee) [inline README.md:126]:
```suggestion
| `--include-[function,macro,struct,union,typedef,var]<String>` | Include a symbol of the given name and kind in the generated bindings (see below). When one of these options is specified, any symbol that is not matched by any specified filters is omitted from the generated bindings. |
```

**Review by JornVernee — APPROVED** [review]
Looks good.

File: README.md (hunk 5 above)
Comment (sundararajana) [reply README.md:126]: looks good

**Pull request closed without merging**